service/history/workflow/update/update.go

```diff
@@ -159,31 +163,86 @@ func newCompleted(
 // WaitOutcome observes this Update's completion, returning the Outcome when it
 // is available. This call will block until the Outcome is known or the provided
 // context.Context expires. It is safe to call this method outside of workflow lock.
-func (u *Update) WaitOutcome(ctx context.Context) (*updatepb.Outcome, error) {
-	return u.outcome.Get(ctx)
+func (u *Update) WaitOutcome(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
+	outcome, err := u.outcome.Get(ctx)
+	if err != nil {
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, outcome, err
+	}
+	return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome, nil
 }
 
 // WaitAccepted blocks on the acceptance of this update, returning nil if has
 // been accepted but not yet completed or the overall Outcome if the update has
 // been completed (including completed by rejection). This call will block until
 // the acceptance occurs or the provided context.Context expires.
 // It is safe to call this method outside of workflow lock.
-func (u *Update) WaitAccepted(ctx context.Context) (*updatepb.Outcome, error) {
+func (u *Update) WaitAccepted(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
 	if u.outcome.Ready() {
```

Comment:
Is there a race here?
Is there any chance that this would return `false` and the `accepted` future doesn't resolve?
I'm not sure I understand this condition, why not always wait on the `accepted` future directly and skip this check?

Reply:
> why not always wait on the accepted future directly and skip this check?

Matt did this so that the Success/Failure outcome of a completed Update can be returned here. If we waited on the `accepted` future we would not have access to them.

> Is there a race here?
> Is there any chance that this would return `false` and the `accepted` future doesn't resolve?

No: the `accepted` future is resolved both on reject and on accept, so (with a correctly implemented SDK) it is impossible for an Update to get to Completed without resolving the `accepted` future.

Reply:
The state machine (ignoring rollbacks) is

<img width="410" alt="image" src="https://github.com/temporalio/temporal/assets/52205/bb444234-ce02-4552-b3dc-7320230c15bb">